feat(deploy): add Source Control Management context to api request

cs-simental opened the pull request:
[CDAAS-2741](https://armory.atlassian.net/browse/CDAAS-2741)

Move the logic for generating the information from the github action to the armory-cli

This change adds a Sim property to the request made to the deploy engine API so that work in Deploy engine will later saved and this information will be available for the PR lineage in the UI.

This was tested by running the armory client pointing to the remote-dev armory address using the new flag --with-scm. Using the goland debugging tools I was able to validate that the scmc data is being add to the request and the pipeline is triggered and deployed without issues


[CDAAS-2741]: https://armory.atlassian.net/browse/CDAAS-2741?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Imported github client

Started the github context retrieval

Commit:
Added scmc to the request

Source Control Management Context added to the deployment request under sourceControl

Commit:
Fixed and issue with scmc marshalling

Changed the ScmContext and made and embedded struct githubScmc

Commit:
Added missing new parameter to start tests

Commit:
Removed unused err value

Commit:
mod and sum changes

mrampton commented on cmd/deploy/start.go:
```diff
@@ -125,6 +128,7 @@ func NewDeployStartCmd(configuration *config.Configuration) *cobra.Command {
 	cmd.Flags().StringVarP(&options.application, "application", "n", "", "application name for deployment")
 	cmd.Flags().StringArrayVarP(&options.targetFilters, "targetFilters", "t", []string{}, "targets specified in the config file to include. Those not specified will be skipped. All specified in the config will be overridden")
 	cmd.Flags().StringToStringVar(&options.context, "add-context", map[string]string{}, "add context values to be used in strategy steps")
+	cmd.Flags().BoolVarP(&options.withScm, "with-scm", "j", false, "add source lineage to be shown in the source card")
```
Let's think about the wording that we use to describe the command/feature and update before merging this PR

Commit:
Removed TODO comment

mrampton commented on cmd/deploy/start.go:
```diff
@@ -158,7 +162,13 @@ func start(cmd *cobra.Command, configuration *config.Configuration, options *dep
 		withConfiguration = WithLocalFile
 	}
-	startResp, rawResp, err = withConfiguration(cmd, options, deployClient)
+	var scmc scm.ScmContext
```
```suggestion

	var scmc scm.ScmContext
```

mrampton commented on cmd/sourceControl/scm.go:
```diff
@@ -0,0 +1,55 @@
+	workflowDispatch Event     = "workflow_dispatch"
+)
+
+type ScmContext interface {
```
```suggestion
type Context interface {
```

mrampton replied:
This is just to resolve the stutter in scm.ScmContext.

mrampton commented on pkg/deploy/request.go:
```diff
@@ -82,6 +85,9 @@ func convertPipelineOptionsToAPIRequest(options StartPipelineOptions) (map[strin
 	}
 	deployment[contextKey] = context
+
+	deployment[scmcKey] = options.Scmc
```
Let's group this with the others (so move to line 79, with a blank line before context := ...)

mrampton commented on cmd/sourceControl/scm.go:
```diff
@@ -0,0 +1,55 @@
+	if err != nil {
+		msg := color.New(color.FgYellow, color.Bold).Sprint("scm error: ")
+		fmt.Fprintf(cmd.OutOrStdout(), "%s %s\n\n", msg, err)
+	}
```
```suggestion
	scmc, err = GithubScmc{}.GetContext()
	if err != nil {
		log.S().Warnf("could not retrieve scm context %v", outLocation, err)
	}
```

mrampton replied:
I think something like this might be more consistent with the rest of the cli

Commit:
Changes to enable and add tests

Commit:
changed with-scm command option description

fieldju commented on cmd/deploy/start.go:
```diff
@@ -238,6 +248,7 @@ func WithLocalFile(cmd *cobra.Command, options *deployStartOptions, deployClient
 		UnstructuredDeployment:  payload,
 		ApplicationNameOverride: options.application,
 		Context:                 options.context,
+		Scmc:                    scmc,
```
```suggestion
		SCMC:                    scmc,
```

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

fieldju replied:
I'm not gonna copy paste this comment everywhere but `Scmc` should be `SCMC` or `scmc` everywhere.

cs-simental replied:
Changed all Scmc references to SCMC

fieldju reviewed (dismissed):
~I'm writing up a details comment about modeling data and will post here soon~

fieldju commented on cmd/sourceControl/github.go:
```diff
@@ -0,0 +1,158 @@
+		Github: GithubData{
+			RunId:    os.Getenv(ghRunId),
+			Workflow: os.Getenv(ghWorkflow),
+		}}
```
```golang
	GithubContext struct {
		BaseContext
		Github  GithubData `json:"github,omitempty"`
		service GithubService
	}
```

Does :point_up: cause this to produce JSON that looks like the following, where base context and GitHub are merged into a single object?

```json
{
  ...
  "scmc": {
    "type": "github",
    ...
    "github": {
      "runId": "foo",
      ...
    }
  }
}
```

cs-simental replied:
Yes, this is an example of the output:
```json
"scm":
    {
        "type": "github",
        "event": "pull_request",
        "reference": "branch",
        "referenceName": "5/merge",
        "source": "pulling",
        "target": "main",
        "principal": "cs-simental",
        "triggeringPrincipal": "cs-simental",
        "prTitle": "test valid token",
        "prUrl": "https://github.com/cs-simental/potato-facts/pull/5",
        "sha": "ebc38b76b1eba3fcb27be7af82064e8a1c988bc",
        "repository": "cs-simental/potato-facts",
        "server": "https://github.com",
        "github":
        {
            "runId": "234568",
            "workflow": "Deploy to CD-as-a-Service"
        }
```

Commit:
Changed Scmc property names to SCMC

Commit:
Fixed SCM test
Taking into account the Env variables already being there when the test are run in github

Commit:
Changed MockServiceProvider

Commit:
Revert file name change

fieldju reviewed (approved):
When you merge this it gets auto-published as a dev release which some (not all though) of our projects use.

So once the actions are complete you can UAT this before promoting the dev release to a proper full release that is consumed by customers

Commit:
Added deploy-engine as a dependency for SCM

Commit:
Fixed TestDeployStatus/yaml_format_success

added missing property 'sourcecontrol: null'

mrampton commented on cmd/sourceControl/github.go:
```diff
@@ -0,0 +1,154 @@
+	ghServer          = "GITHUB_SERVER_URL"
+	ghRunId           = "GITHUB_RUN_ID"
+	ghWorkflow        = "GITHUB_WORKFLOW"
+)
```
These should all be available in deploy-engine:api -- but the names start as Github (so GithubToken, etc).

mrampton commented on cmd/sourceControl/scm.go:
```diff
@@ -0,0 +1,51 @@
+	pullRequest      de.Event     = "pull_request"
+	push             de.Event     = "push"
+	workflowDispatch de.Event     = "workflow_dispatch"
+)
```
These are likewise available in deploy-engine :)

Commit:
Changed scm parameter to sourceControl

Commit:
Added validations to scm

Validate that env variables are not empty
Validate that pr pointer is not empty when a bad token is given

Commit:
Added one test to scm

Commit:
Removed duplicated constants
